File: python/stencil-bench.py
#!/usr/bin/env python3

import time
import sys

def traverse(order: str, multiplier: int):
    """
    Create a square array, fill it with data then compute an averaging stencil
    over it in either row-major or column-major order.

    `order` must be one of `row` or `column`.
    """

    n = 1000 * multiplier
    x = [[i*n+j for j in range(0, n)] for i in range(0, n)]
    if order == "column":
        for i in range(1, n-1):
            for j in range(1, n-1):
                x[j][i] = (x[j][i]+x[j-1][i]+x[j+1][i]+x[j][i-1]+x[j][i+1]) / 5.0
    elif order == "row":
        for i in range(1, n-1):
            for j in range(1, n-1):
                x[i][j] = (x[i][j]+x[i-1][j]+x[i+1][j]+x[i][j-1]+x[i][j+1]) / 5.0
    else:
        print(f"traverse received bad order: must be \"column\" or \"row\", got \"{order}\"")

def time_traverse(order: str, multiplier: int):
    t1 = time.time()
    traverse(order, multiplier)
    t2 = time.time()
    t = t2-t1
    t_ms = t*1000
    # TODO 2023-01-30 raehik: RebenchLogAdapter doesn't support fractional
    # components! Integers only. No good reason, should be changing this soon.
    t_ms_int = int(t_ms)
    print(format_rebenchlog_line("array_stencil", 1, t_ms_int))

def format_rebenchlog_line(name, iters, runtime):
    return f"{name}: iterations={iters} runtime: {runtime}ms"

def main():
    argc = len(sys.argv)-1
    if argc != 2:
        print(f"error: expected exactly 2 arguments, got {argc}")
        sys.exit(1)
    multiplier = int(sys.argv[2])
    if sys.argv[1] == "-r":
        for i in range(0, 5):
            time_traverse("row", multiplier)
    elif sys.argv[1] == "-c":
        for i in range(0, 5):
            time_traverse("column", multiplier)
    else:
        print("error: expected -r or -c option")

main()

File: fortran/stencil-bench.f90
program example
  implicit none

  real              :: startTime, finishTime
  character(len=32) :: argument1
  character(len=32) :: argument2
  integer           :: multiplier  = 1
  logical           :: colMajorFlag = .true.


  ! configuration of this benchmark from command line args
  
  if (command_argument_count() > 1) then

    ! size parameter
    if (command_argument_count() > 2) then
      call get_command_argument(3, argument2)
      read (argument2,'(I10)') multiplier
      print '("Size = ",i10)', multiplier
    end if

    call get_command_argument(2, argument1)

    ! process arguments
    select case (argument1)
      case ('-c', '--column')
        colMajorFlag = .true.
        print '("Column-major mode.")'
        call benchmark()
        stop

      case ('-r', '--row')
        colMajorFlag = .false.
        print '("Row-major mode.")'
        call benchmark()
        stop

      case default
        call print_help()
        stop
      end select

  else
    call print_help()
  end if

contains

    subroutine benchmark()
      call cpu_time(startTime)
      ! code to test
      call traverse(colMajorFlag, multiplier)
      call cpu_time(finishTime)
      ! report results for PlainSecondsLog
      print '(f6.3)', (finishTime - startTime)
    end subroutine benchmark

    ! Create a square array, fill it with data
    ! then compute an averaging stencil over it
    ! in either row-major or column-major order
    subroutine traverse(colMajorFlag, multiplier)
      logical                :: colMajorFlag
      integer                :: multiplier
      integer                :: n
      real, dimension (:, :), allocatable :: x
      integer                :: i, j

      n = 10000 * multiplier
      allocate (x(n, n))

      ! Fill up array with some data
      do i = 1, n
         do j = 1, n
            x(i,j) = i*n+j
         end do
      end do

      ! Do convolution (with flag to flip which dimension
      ! gets 'visited' first).
      do i = 2, n-1
         do j = 2, n-1
            if (colMajorFlag) then
               x(j,i) = (x(j,i)+x(j-1,i)+x(j+1,i)+x(j,i-1)+x(j,i+1))/5.0
            else
               x(i,j) = (x(i,j)+x(i-1,j)+x(i+1,j)+x(i,j-1)+x(i,j+1))/5.0
            end if
         end do
      end do

      deallocate (x)

    end subroutine traverse

    ! Help routine
    subroutine print_help()
        print '(a, /)', 'example benchmark <option> [multiplier]'
        print '(a, /)', 'command-line options:'
        print '(a)',    '  -c, --column      run benchmark in column-major mode'
        print '(a)',    '  -r, --row         run benchmark in row-major mode'
        print '(a, /)', '  -h, --help        print usage information and exit'
        print '(a, /)', 'e.g., example benchmark -r 10'
    end subroutine print_help

end program example
